### ISSUE TITLE
New lint: manual_memmove

### ISSUE THREAD

**Issue by untitaker** [rust-lang/rust-clippy#8335]
### What it does

Like manual_memcpy, find for-loops equivalent to `slice::copy_within`.

### Lint Name

manual_memmove

### Category

perf

### Advantage

It's shorter and faster.

### Drawbacks

any case where memmove is not actually faster? (should probably file as bug against rust though)

### Example

```rust
let mut arr: [u8; 4] = [1, 2, 3, 4];
                                     
for i in 0..2 {
    arr[i] = arr[i + 2];
}

```

Could be written as:

```rust
arr.copy_within(2..4, 0);
```

**Comment by untitaker** [comment]
I implemented a mvp of this in #8337.
